onnx/src/ops/ml/tree.rs

```diff
@@ -0,0 +1,664 @@
+    fn aggregate(&mut self, score: f32, total: &mut f32) {
+        *total += score;
+        self.count += 1;
+    }
+
+    fn post_aggregate(&mut self, total: &mut f32) {
+        if self.count > 1 {
+            *total /= (self.count as f32);
+        }
+        self.count = 0;
+    }
+}
+
+#[derive(Clone, Copy, Default, Debug)]
+pub struct MaxFn;
+
+impl AggregateFn for MaxFn {
+    fn aggregate(&mut self, score: f32, total: &mut f32) {
+        *total = total.max(score);
+    }
+}
+
+#[derive(Clone, Copy, Default, Debug)]
+pub struct MinFn;
+
+impl AggregateFn for MinFn {
+    fn aggregate(&mut self, score: f32, total: &mut f32) {
+        *total = total.min(score);
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Aggregate {
+    Sum,
+    Avg,
+    Max,
+    Min,
+}
+
+impl Default for Aggregate {
+    fn default() -> Self {
+        Aggregate::Sum
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum PostTransform {
+    Softmax,
+    Logistic,
+    SoftmaxZero,
+    // Probit, // probit, especially multinomial, is p.i.t.a. - so let's ignore it for now
+}
+
+impl PostTransform {
+    pub fn apply(&self, scores: &mut ArrayViewMut1<f32>) {
+        match self {
+            PostTransform::Softmax => {
+                // TODO: stability w.r.t. nan/inf/-inf?
+                let max: f32 = scores
+                    .iter()
+                    .cloned()
+                    .max_by(|a, b| a.partial_cmp(&b).unwrap_or(Ordering::Equal))
+                    .unwrap_or(0.0);
+                let mut norm = 0.;
+                scores.map_inplace(|a| {
+                    *a = (*a - max).exp();
+                    norm += *a;
+                });
+                scores.map_inplace(|a| *a = *a / norm);
+            }
+            PostTransform::Logistic => {
+                scores.map_inplace(|a| {
+                    let v = 1. / (1. + f32::exp(-f32::abs(*a)));
+                    *a = if a.is_sign_negative() { 1. - v } else { v };
+                });
+            }
+            PostTransform::SoftmaxZero => {
+                // TODO: stability w.r.t. nan/inf/-inf?
+                let max: f32 = scores
+                    .iter()
+                    .cloned()
+                    .filter(|&a| a != 0.)
+                    .max_by(|a, b| a.partial_cmp(&b).unwrap_or(Ordering::Equal))
+                    .unwrap_or(0.0);
+                let mut norm = 0.;
+                scores.map_inplace(|a| {
+                    if *a != 0. {
+                        *a = (*a - max).exp();
+                        norm += *a;
+                    }
+                });
+                scores.map_inplace(|a| *a = *a / norm);
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Tree {
+    n_classes: usize,
+    nodes: Vec<Node>, // TODO: can this be a slice/view into ensemble's contiguous storage?
+    leaves: Vec<Leaf>, // TODO: store as a collection of direct slices instead of indices?
```

Comment:
I think indices in vectors are a good compromise. Making trees work with slices make the code very difficult, if even possible, as it leads to self borrowing structures.

 If we have to worry about performance here, we can use unchecked vector index (it's unsafe, but we can live with it as the structure is fixed).

Reply:
Oh, I didn't mean self-borrowing. More like - the ensemble owns a contiguous collection of nodes / leaves, and trees just receive slices into that. However then it wouldn't be possible to just simply create a new tree without first creating an ensemble and then indexing into it, also there would be lifetimes all over the place... things would get a bit weird, so not sure it would make things a lot better.

For a generic solution, I think this is ok. I had ideas to add specializations later for near-balanced binary trees of the most common depths (e.g. up to 16; I've rarely seen more), which would be lightning fast barely with any lookups etc.. we'll handle it later, that's the fun part ;)